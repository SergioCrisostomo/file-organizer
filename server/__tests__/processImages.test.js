const path = require("path");
const getFilesFromSelection = require("../utils/getFilesFromSelection");
const findDuplicateFolders = require('../utils/findDuplicateFolders');

const filesAnalyser = require("../../lib/index");
const testImagesFolder = path.join(__dirname, "/images");
const projectRoot = path.join(__dirname, "../../../");

const getTestFiles = async () => {
  const fetchers = getFilesFromSelection([testImagesFolder], [".jpg"]);
  const paths = await Promise.all(fetchers);
  return paths.flat();
};

const anonymizeFolders = (str) => {
  return str.split(projectRoot).join("");
};

test("Detect correct amount of images", async () => {
  const files = await getTestFiles();
  expect(files.length).toBe(7);
});

test("Generate a report", async () => {
  const files = await getTestFiles();
  const report = await filesAnalyser(files);
  const expectedReport = [
    [
      "file-organizer/server/__tests__/images/folder_a/bairro_alto-elevador_da_bica.jpg",
      "file-organizer/server/__tests__/images/folder_b/bairro_alto-elevador_da_bica.jpg",
      "file-organizer/server/__tests__/images/folder_b/bairro_alto-elevador_da_bica0.jpg",
    ],
    [
      "file-organizer/server/__tests__/images/folder_a/beach.jpg",
      "file-organizer/server/__tests__/images/folder_b/beach.jpg",
    ],
  ];
  const input = anonymizeFolders(JSON.stringify(report));
  const output = anonymizeFolders(JSON.stringify(expectedReport));
  expect(input).toEqual(output);
});

test("Should get file stats", async () => {
  const testFile = path.join(
    projectRoot,
    "/file-organizer/server/__tests__/images/folder_a/bairro_alto-elevador_da_bica.jpg"
  );
  const getFileStats = require("../../server/routes/getFileStats");
  let response = null;
  await getFileStats(
    { query: { file: testFile } },
    {
      send(data) {
        response = data;
      },
    }
  );

  // remove access timestamps, not relevant to the test...
  ["atime", "atimeMs"].forEach((key) => delete response.stats[key]);

  expect(response).toMatchSnapshot();
});

test("Should compare folders", async () => {
  const files = await getTestFiles();
  const duplicates = await filesAnalyser(files);
  const duplicateFolders = await findDuplicateFolders(duplicates);

  expect(duplicateFolders).toMatchSnapshot();
});
