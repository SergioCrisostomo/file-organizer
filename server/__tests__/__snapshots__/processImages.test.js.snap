// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should compare folders 1`] = `
Array [
  Object {
    "A": "/Users/sergiocrisostomo/github/file-organizer/server/__tests__/images/folder_a",
    "B": "/Users/sergiocrisostomo/github/file-organizer/server/__tests__/images/folder_b",
    "brokenLinks": Object {
      "distinctBrokenLinks": 0,
      "leftBrokenLinks": 0,
      "rightBrokenLinks": 0,
      "totalBrokenLinks": 0,
    },
    "differences": 2,
    "differencesDirs": 0,
    "differencesFiles": 2,
    "distinct": 1,
    "distinctDirs": 0,
    "distinctFiles": 1,
    "equal": 2,
    "equalDirs": 0,
    "equalFiles": 2,
    "left": 0,
    "leftDirs": 0,
    "leftFiles": 0,
    "right": 1,
    "rightDirs": 0,
    "rightFiles": 1,
    "same": false,
    "symlinks": undefined,
    "total": 4,
    "totalDirs": 0,
    "totalFiles": 4,
  },
]
`;

exports[`Should get file stats 1`] = `
Object {
  "metadata": Object {
    "DateTime": "2017:01:24 18:49:24",
    "ExifIFDPointer": 196,
    "Make": "Canon",
    "Model": "Canon EOS 1000D",
    "Orientation": 1,
    "ResolutionUnit": 2,
    "SubExif": Object {
      "ApertureValue": Array [
        4,
      ],
      "ColorSpace": 1,
      "ComponentsConfiguration": "0x01020300",
      "CustomRendered": 0,
      "DateTimeDigitized": "2017:01:24 18:49:24",
      "DateTimeOriginal": "2017:01:24 18:49:24",
      "ExifVersion": "0221",
      "ExposureBiasValue": 0,
      "ExposureMode": 0,
      "ExposureProgram": 2,
      "ExposureTime": Array [
        0.003125,
      ],
      "FNumber": Array [
        4,
      ],
      "Flash": 16,
      "FlashpixVersion": "0x30313030",
      "FocalLength": Array [
        50,
      ],
      "FocalPlaneResolutionUnit": 2,
      "FocalPlaneXResolution": Array [
        4438.356164383562,
      ],
      "FocalPlaneYResolution": Array [
        4445.969125214408,
      ],
      "InteroperabilityIFDPointer": 9478,
      "MakerNote": "0x2000010003002f0000001c04000002",
      "MeteringMode": 5,
      "PhotographicSensitivity": 200,
      "PixelXDimension": 3888,
      "PixelYDimension": 2592,
      "SceneCaptureType": 0,
      "ShutterSpeedValue": 8.375,
      "SubSecTime": "53",
      "SubSecTimeDigitized": "53",
      "SubSecTimeOriginal": "53",
      "UserComment": "0x000000000000000000000000000000",
      "WhiteBalance": 0,
    },
    "XResolution": Array [
      72,
    ],
    "YCbCrPositioning": 2,
    "YResolution": Array [
      72,
    ],
  },
  "stats": Stats {
    "birthtime": 2020-05-21T01:17:20.000Z,
    "birthtimeMs": 1590023840000,
    "blksize": 4096,
    "blocks": 736,
    "ctime": 2020-05-23T08:09:15.473Z,
    "ctimeMs": 1590221355472.521,
    "dev": 16777220,
    "gid": 20,
    "ino": 12896680906,
    "mode": 33188,
    "mtime": 2020-05-21T01:17:20.000Z,
    "mtimeMs": 1590023840000,
    "nlink": 1,
    "rdev": 0,
    "size": 372781,
    "uid": 501,
  },
}
`;
